Workspaces copy/move support

joshhaug opened the pull request:
* **Tickets addressed:** No ticket
* **Review:** By file
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
At a high level, this just allows the user to specify one of two POSTs:

```json
{
  "moveTo": "moved/file/location.txt",
  "toWorkspace": 1
}
```

```json
{
  "copyTo": "copied/file/location.txt",
  "toWorkspace": 1
}
```

If `toWorkspace` is not specified, or is the same as the context's `PathInformation.workspaceId`, then these are assumed to be operations within the given context workspace. Otherwise, they are performed across workspaces. 

The `post` method now calls out to `handleCopy` or `handleMove`, which perform some shared tests via `isCopyOrMoveValid`. If those pass, we call the appropriate `(move|copy)(Directory|File)` method. These methods accept both a source and target workspace Id, which prevents duplicated logic between intra- and inter-workspace copy/move ops. 

Additionally, I fixed a typo in `GetExtensionMappingsAction` which was preventing things from working when testing.

joshhaug committed:
Prototype of moving files and directories from one workspace to another

joshhaug committed:
Added copyFile functionality, and refactored to use one method for intra- and inter-workspace copy/move operations

joshhaug committed:
Moving common checks in copy/move ops to be in a utility function

joshhaug committed:
Cleaned up some todos, etc.

joshhaug committed:
Adding additional errors to workspace copy/move operations

joshhaug committed:
Adding file-based workspace config stuff to action-server

joshhaug committed:
Not allowing directory overwrite when handing put operations

joshhaug committed:
Updating codeRunner.ts to use simplified auth

joshhaug committed:
Refactor of placement of WorkspaceFileOpException, and cleanup in codeRunner.ts

joshhaug committed:
Fixed issue with filename not being equal to uploaded file name

joshhaug committed:
Cleanup of isCopyMoveValid function

joshhaug committed:
Moving action-server changes to new branch

joshhaug committed:
Whitespace cleanup

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -213,25 +215,27 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
                     .allowNullable()
                     .check(Objects::nonNull, "'type' must be provided.")
-                    .check(ts -> ts.equalsIgnoreCase("file") || ts.equalsIgnoreCase("directory"),
+                    .check(ts -> ts == null || ts.equalsIgnoreCase("file") || ts.equalsIgnoreCase("directory"),
```
`ts` physically cannot be null, see L217 `.check(Objects::nonNull, "'type' must be provided.")`. Please remove this change or explain why it is necessary.

joshhaug replied:
I included this because this was throwing a failure when performing the `equalsIgnoreCase` check. I'm confused by this as presumably the above `check()` would short-circuit, but I guess the behavior of this is a bit more nuanced than I thought.

Mythicaeda replied:
Well, either way, `null` is **not** a permitted value. If an exception is occurring when `ts` is `null`, why not just reverse the check instead?

```suggestion
                    .check(ts -> "file".equalsIgnoreCase(ts) || "directory".equalsIgnoreCase(ts),
```

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -213,25 +215,27 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
       overwrite = context.queryParamAsClass("overwrite", Boolean.class).getOrDefault(false);
     } catch (ValidationException ve) {
-      context.status(400).result(ve.getMessage());
+      context.status(400).result(ve.getMessage() != null ? ve.getMessage() : "Invalid request");
```
Can you please provide an example case where `ve.getMessage` was `null`? A thrown exception should always have a `message` AFAIK

joshhaug replied:
I encountered this during debugging but I don't remember the case that triggered it. I was surprised as well. I can try to dig up the case that triggered this -- it could've potentially been a cascade from another bug.

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -213,25 +215,27 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
+    // Report a "Conflict" status if the file already exists, is a file and not a directory, and "overwrite" is false
+    if(workspaceService.checkFileExists(pathInfo.workspaceId, pathInfo.filePath)
+       && !workspaceService.isDirectory(pathInfo.workspaceId, pathInfo.filePath)
+       && !overwrite) {
```
**Major**: `isDirectory` should **not** be added here. This is a check for if a resource with that name already exists in **any** form (as indicated by it returning a [409: Conflict](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/409)). More critically, this check is happening **BEFORE** we've even distinguished if the user is trying to create a `file` or a `directory`.

You can add a "make sure there's not a directory with this name" check within the file-uploading part of the method (between lines 235 and 241) if you'd like, but please revert the change here.

joshhaug replied:
Done in e86d0c325be6a100306a5811660b431da9ff6cfb

Mythicaeda replied:
Comment on L227 was not reverted.

joshhaug replied:
Good catch, updated in 5b3b18db8f72fc8e905ec6d95846ea80460d9ac0

Mythicaeda replied:
Since I know that I'm going to be editing this part of the code after we remerge, I'll call this comment resolved, but you ought to know that this is still not the original message. The original message was `Report a "Conflict" status if the object already exists and "overwrite" is false`. In the future, I'd recommend dropping or editing prior commits to rollback a change in order to avoid subtle differences like this.

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -213,25 +215,27 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
       // Reject the request if the file isn't provided.
       final var file = context.uploadedFile("file");
-      if (file == null || !pathInfo.fileName().equals(file.filename())) {
+      if (file == null) {
```
Please revert the change here. The second half of this `if` statement ensures that the request was formatted correctly. That is, that the request said "I'm uploading file `XYZ.seq`, here is the file `XYZ.seq`"

joshhaug replied:
My opinion is that `file.filename()` may not be the same as `pathInfo.fileName()`. For the following action code, for instance:
```ts
  async writeFile(
    name: string,
    contents: string,
    overwrite: boolean = false
  ): Promise<any> {
    const formData = new FormData();
    formData.append("file", new Blob([contents]), name);
    const path = `/ws/${this.workspaceId}/${encodeURIComponent(name)}?type=file&overwrite=${overwrite}`;
    await this.reqWorkspace(
      path,
      'PUT',
      formData
    );
    return { success: true };
  }
```
This check could fail if the `name` is `dir1/dir2/test.json`, because `pathInfo.fileName()` would be `test.json` but `file.filename()` would be `dir1/dir2/test.json`.  Do we consider this to be a incorrectly formatted request?

Mythicaeda replied:
`file.filename()` would be `test.json`. That's the explicit implementation of the method in `PathInfo` (L42 of this file)

joshhaug replied:
`file.filename()` comes from `context.uploadedFile("file")`, which is set via the FormData above. From my testing it seems like `filename` is the full path of the file (`name` in the above typescript example).

Mythicaeda replied:
Yes, but that TypeScript request is misformatted.

This line:
`formData.append("file", new Blob([contents]), name);`
should set the file name to just the actual file name, not the complete path.

joshhaug replied:
Addressed in c479eb806b63fcd44bf30e259e160cc5a8856591

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceService.java:
```diff
@@ -47,15 +47,52 @@ record FileStream(InputStream readingStream, String fileName, long fileSize){}
    */
-  boolean moveFile(final int workspaceId, final Path oldFilePath, final Path newFilePath)
-  throws NoSuchWorkspaceException, SQLException;
+  boolean moveFile(final int oldWorkspaceId, final Path oldFilePath, final int newWorkspaceId, final Path newFilePath)
```
Open for discussion: I think the "between workspaces" case should be an override of `moveFile`/`moveDirectory`, as it saves us a needless DB call in the far more common case of moving files/directories within a workspace.

joshhaug replied:
Would you mind clarifying? I'm not sure I follow 😕

Mythicaeda replied:
I'm proposing having the following overrides of `moveFile`:
```java
boolean moveFile(final int workspaceId, final Path oldFilePath, final Path newFilePath)

boolean moveFile(final int oldWorkspaceId, final Path oldFilePath, final int newWorkspaceId, final Path newFilePath)
```

joshhaug replied:
Implemented an override in 5b3b18db8f72fc8e905ec6d95846ea80460d9ac0, but it is not currently being used.

Mythicaeda replied:
Right. I was proposing creating _and using_ it. If it's going to go unused, then I'd rather the change be reverted/dropped.

dandelany replied:
Fixed in [876e7ec](https://github.com/NASA-AMMOS/aerie/pull/1712/commits/876e7ec47e930e5747528e90a3070cb5ccb23798) - discussed with @mattdailis & we decided to remove the override and add a check in the `moveFile` implementation to avoid the extra DB check if the IDs are the same.

Mythicaeda replied:
Thank you. You should do the same for `moveDirectory` and `copyFile/copyDirectory` so that the behavior isn't unique to `moveFile`

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceService.java:
```diff
@@ -47,15 +47,52 @@ record FileStream(InputStream readingStream, String fileName, long fileSize){}
+   * @param newDirectoryPath the new path of the directory, relative to the new workspace root
+   * @return true if the directory was moved, false otherwise
+   */
+  boolean moveDirectory(final int oldWorkspaceId, final Path oldDirectoryPath, final int newWorkspaceId, final Path newDirectoryPath)
```
Minor: Can you please move `moveDirectory`/`copyDirectory` back to where it was in the file? Prior to this change, the class was organized to have all of the file methods grouped together, and all of the directory methods grouped together, both groups ordered as `RCUD` (which, to be fair, I probably should've ordered them in the more standard `CRUD` order, but still). Putting the directory methods next to the file methods breaks this organization.

joshhaug replied:
Done in c5505dfaa7b4262dba069c95358ff92dedcb795d

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java:
```diff
@@ -195,17 +281,22 @@ public boolean createDirectory(final int workspaceId, final Path directoryPath)
-    if(Files.isSameFile(oldPath, repoPath) || Files.isSameFile(newPath, repoPath)) return false;
+    if(Files.isSameFile(oldPath, oldRepoPath)) throw new WorkspaceFileOpException("Cannot move the workspace root directory.");
+
+    // Do not permit a directory to replace the root directory
+    // TODO: maybe we want to allow this? This check fails anyways because newPath is not a real file at this point
+    // if(Files.isSameFile(newPath, newRepoPath)) return false;
```
Three things:
1) `TODO` is left behind
2) `newPath` should absolutely be a real file if it's the other repo's root path, not sure what that you mean by that part of the `TODO`. Could you elaborate?
3) No, we don't want to allow the user to replace the root directory of the new workspace (this is to maintain consistency with the "within workspace" interaction)

joshhaug replied:
Done in ed2df9329d43e046920f89024c07ad4145ca193b

joshhaug committed:
Removed directory case in conflict check

joshhaug committed:
Removed TODO and added check for replacing root directory

joshhaug committed:
Moved copyDirectory and moveDirectory

joshhaug committed:
Adding check back in for filenames being identical

joshhaug committed:
Update workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java

Co-authored-by: Mythicaeda <theresa.kamerman@jpl.nasa.gov>

joshhaug commented:
Just reviewed these comments with @dandelany , I think we are ready for re-review if you have some time @Mythicaeda

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -252,58 +255,137 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
+    } catch (Exception e) {
+      e.printStackTrace();
+      context.status(400).result("Error: " + e.getMessage() != null ? e.getMessage() : "Invalid request");
+    }
```
Javalin already has a generic exception handler. Please replace this block with the actual exceptions you expect to catch (and update the method signature accordingly). Additionally, status code 400 is **not** appropriate as a catch-all error code. `400` means that there was something wrong that in the client's request (ie, their body was misformatted, they forgot a query variable)

joshhaug replied:
Updated in a14d93e5202b0401531dd381dbaf86cb7b92b417

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -252,58 +255,137 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
 
-  private void post(Context context) throws NoSuchWorkspaceException, IOException {
-    final var pathInfo = PathInformation.of(context);
+  private void post(Context context) throws IOException, SQLException, NoSuchWorkspaceException {
```
The `post` method used to contain a `helpText` and would respond with the appropriate body format should the user's body be misformatted. The removal of this behavior is a regression and makes the API harder to use for clients who do not already know the appropriate format. Please restore and expand on the helpText.

joshhaug replied:
Good idea, updated in 05b35926a70b1061edc8d6461856fd97eb7fa2c7

joshhaug commented:
Note to myself: need to check/verify empty file loading

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -252,58 +255,137 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
-        if (!workspaceService.checkFileExists(pathInfo.workspaceId, pathInfo.filePath)) {
-          context.status(404).result(pathInfo.fileName() + " does not exist in the workspace.");
-          return;
+    return new CopyMoveValid(true, 0, "");
```
Since you're using HTTP status codes as the values elsewhere, you should set status to `200` here. Additionally, you don't need the boolean `valid`, since `status` already communicates that.

joshhaug replied:
Ok, updated in 6312c42f1a2dc5f97ad00945021a0c3d38851c93

joshhaug commented:
> Note to myself: need to check/verify empty file loading

Fix implemented in aerie-actions library, and did not require any changes in backend as I originally thought: https://github.com/NASA-AMMOS/aerie-actions/pull/8#issuecomment-3115104197

joshhaug committed:
Updated comment in workspace bindings, and added override for moveFile method

joshhaug committed:
Updated CopyMoveValid to not include a status

joshhaug committed:
Updated caught conditions in post

joshhaug committed:
Added helpText and cleaned up error messages a bit more for post

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -252,58 +257,172 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
+
+    } catch (IOException | SQLException e) {
+      // Internal server error
+      e.printStackTrace();
```
Occurring in multiple locations and multiple files: The correct way for the server to log an error is by using a `org.slf4j.Logger`. Please update this to make Workspaces consistent with the rest of our services

Specifically, this occurs twice in `WorkspaceBindings` and twice in `WorkspaceFileSystemService`.

dandelany replied:
Fixed in ab8d781

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java:
```diff
@@ -195,17 +288,21 @@ public boolean createDirectory(final int workspaceId, final Path directoryPath)
+    if(Files.isSameFile(oldPath, oldRepoPath)) throw new WorkspaceFileOpException("Cannot move the workspace root directory.");
+
+    // Do not permit a directory to replace the root directory
+    if(Files.isSameFile(newPath, oldRepoPath)) throw new WorkspaceFileOpException("Cannot replace the workspace root directory.");
```
You are missing a check to prevent replacing the root directory of the other workspace:

```
Files.isSameFile(newPath, newRepoPath)
```

dandelany replied:
* Fixed in ab8d781
* I think we *only* care about this check, not `oldRepoPath`, so I replaced line 305
* `Files.isSameFile` throws if either path doesn't exist, which `newPath` won't in the nominal case, so I made it `Files.exists(newPath) && Files.isSameFile(newPath, newRepoPath)`

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java:
```diff
@@ -159,6 +160,98 @@ public boolean moveFile(final int workspaceId, final Path oldFilePath, final Pat
+    try {
+      // Validate source exists and is a directory
+      if (!Files.exists(sourcePath)) throw new WorkspaceFileOpException("Source directory \"%s\" in workspace %d does not exist.".formatted(sourceFilePath, sourceWorkspaceId));
+      if (!Files.isDirectory(sourcePath)) throw new WorkspaceFileOpException("Source directory \"%s\" in workspace %d is not actually a directory. This is a bug.".formatted(sourceFilePath, sourceWorkspaceId));
```
Please exclude the phrase "This is a bug" from the error message.
```suggestion
      if (!Files.isDirectory(sourcePath)) throw new WorkspaceFileOpException("Source directory \"%s\" in workspace %d is not actually a directory.".formatted(sourceFilePath, sourceWorkspaceId));
```

Additionally, this state is not unsalvageable -> you could just go into the `copyFile` with the same args (However, this "additionally" part is nonblocking, I'm just musing on what can be done instead of throwing)

dandelany replied:
Fixed in ab8d781, I think it's OK to throw in this "shouldn't happen" case so we can figure out why it did

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java:
```diff
@@ -159,6 +160,98 @@ public boolean moveFile(final int workspaceId, final Path oldFilePath, final Pat
+            } else {
+              Files.copy(source, target);
+            }
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
```
If you're not going to handle the IOException, you should let it be thrown as opposed to rewrapping it as Unchecked (especially since you catch `IOException` further on in the method).

dandelany replied:
Annoyingly Java doesn't let you throw checked exceptions like `IOException` unless you declare them, which you can’t do with an inline lambda. So it seems like you could either extract this into a separate function, or throw something else like a `RuntimeException`, but that's even less descriptive. Are either of those much of an improvement/more idiomatic than just doing this? `¯\_(ツ)_/¯`

Mythicaeda replied:
No, if the exception is being wrapped in `Unchecked` to get around lambda restrictions, then `UncheckedIOException` makes the most sense.

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -252,58 +257,172 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
+    } catch (IOException | SQLException e) {
+      // Internal server error
+      e.printStackTrace();
+      context.status(500).result("Internal server error while processing the request.");
```
```suggestion
      context.status(500).result("Internal server error while processing the request." + e.getMessage());
```

dandelany replied:
fixed in ab8d781694b733db717e63c53a2116bf87ea553d

Mythicaeda commented on workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceBindings.java:
```diff
@@ -252,58 +257,172 @@ private void put(Context context) throws NoSuchWorkspaceException, IOException {
+
+  private CopyMoveValid isCopyOrMoveValid(int sourceWorkspace, Path sourceFile, int targetWorkspace, Path targetFile)
+  throws NoSuchWorkspaceException {
+    // Reject if source does not exist
+    if (!workspaceService.checkFileExists(sourceWorkspace, sourceFile)) {
+      return new CopyMoveValid(404, sourceFile + " does not exist in the source workspace.");
+    }
+
+    // Reject if target workspace does not exist
+    if (!workspaceService.checkFileExists(targetWorkspace, Path.of("/"))) {
+      return new CopyMoveValid(404, "Target workspace with ID "+targetWorkspace+" does not exist.");
+    }
```
Nonblocking: I've mixed feelings on keeping these checks, as they're already checked by the actual move/copy procedure, and quite quickly too (especially checking if the other workspace exists). I don't feel strongly about removing them, however, so feel free to resolve this comment.

dandelany replied:
Notably this doesn't seem to be the case in the `moveDirectory` function, which only checks for root directory stuff, but not existence/replacement of existing directories. I would be in favor of only checking once, if we can audit & guarantee they're really checked everywhere they need to be. But like you said, nonblocking.

dandelany committed:
fixes for improved logging and error handling in workspace service

dandelany committed:
removed unused override of moveFile, added ternary to skip extra DB call

Mythicaeda committed:
Use "catches" to catch when Workspaces do not exist

Before this, the if statement that was supposed to validate if the workspace existed would throw a NoSuchWorkspaceException in the event that the workspace didn't exist

Mythicaeda committed:
Remove unneeded DB calls when moving/copying within the same workspace

dandelany commented:
finished review with @Mythicaeda and [tests pass](https://github.com/NASA-AMMOS/aerie/actions/runs/16605212133/job/46976680437) - will resolve the Publish scan issue separately. merging this.
